tests/msc2716_test.go

```diff
@@ -1211,3 +1254,159 @@ func batchSendHistoricalMessages(
+
+	if validateState {
+		// Get the pagination token for the end of the historical batch itself
+		contextRes := c.MustDoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", roomID, "context", expectedEventIDOrder[0]}, client.WithContentType("application/json"), client.WithQueries(url.Values{
+			"limit": []string{"0"},
+		}))
+		contextResBody := client.ParseJSON(t, contextRes)
+		batchStartPaginationToken := client.GetJSONFieldStr(t, contextResBody, "end")
+
+		// Fetch a chunk of `/messages` which only contains the historical batch. We
+		// want to do this because `/messages` only returns the state for the first
+		// message in the `chunk` and we want to be able assert that the historical
+		// state is able to be resolved.
+		messagesRes := c.MustDoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", roomID, "messages"}, client.WithContentType("application/json"), client.WithQueries(url.Values{
+			"dir": []string{"b"},
+			// From the end of the historical batch (this will be pointing at the )
+			"from": []string{batchStartPaginationToken},
+			// We are aiming to scrollback to the start of the existing historical
+			// messages
+			"limit": []string{fmt.Sprintf("%d", len(expectedEventIDOrder))},
+			// We add these options to the filter so we get member events in the state field
+			"filter": []string{"{\"lazy_load_members\":true,\"include_redundant_members\":true}"},
+		}))
+
+		must.MatchResponse(t, messagesRes, match.HTTPResponse{
+			JSON: []match.JSON{
+				// Double-check that we're in the right place of scrollback
+				matcherJSONEventIDArrayInOrder("chunk",
+					expectedEventIDOrder,
+					historicalEventFilter,
+				),
+				// Make sure the historical m.room.member join state event resolves
+				// for the given chunk of messages in scrollback. The member event
+				// will include the displayname and avatar.
+				match.JSONCheckOffAllowUnwanted("state", makeInterfaceSlice(stateEventIDs), func(r gjson.Result) interface{} {
+					return r.Get("event_id").Str
+				}, nil),
+			},
+		})
+	}
+
+	// Make sure the historical events appear in scrollback without jumping back
+	// in time specifically.
+	fullMessagesRes := c.MustDoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", roomID, "messages"}, client.WithContentType("application/json"), client.WithQueries(url.Values{
+		"dir":   []string{"b"},
+		"limit": []string{"100"},
+	}))
+	must.MatchResponse(t, fullMessagesRes, match.HTTPResponse{
+		JSON: []match.JSON{
+			matcherJSONEventIDArrayInOrder("chunk",
+				expectedEventIDOrder,
+				historicalEventFilter,
+			),
+		},
+	})
+}
+
+// Looks through a list of events to find the sliding window of expected event
+// ID's somewhere in the list in order. The expected list can start anywhere in
+// the overall list.
+func matcherJSONEventIDArrayInOrder(wantKey string, expectedEventIDOrder []string, eventFilter func(gjson.Result) bool) match.JSON {
```

Comment:
Please document your inputs more. `wantKey` feels like it should be mandatory (and the term "want" is usually paired with "got" for testing but that isn't the case here) but it's in fact an optional key which is where to start looking for events from. I would rename this to `jsonArrayKey`.